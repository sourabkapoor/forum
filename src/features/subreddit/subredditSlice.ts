import { createSlice } from "@reduxjs/toolkit";
import { DocumentData } from "firebase/firestore";
import { RootState } from "../../redux/store";

interface AuthState {
  communityModalState: boolean;
  communityData: DocumentData;
  userCommunities: DocumentData[];
}

const initialState: AuthState = {
  communityModalState: false,
  communityData: {},
  userCommunities: [],
};

export const selectCommunityData = (state: RootState) =>
  state.subreddit.communityData;

export const selectCommunityModalState = (state: RootState) =>
  state.subreddit.communityModalState;

export const selectUserCommunitiesData = (state: RootState) =>
  state.subreddit.userCommunities;

// export const selectCurrentSubredditData = (state: RootState) =>
//   state.subreddit.currentSubredditData;

const subredditSlice = createSlice({
  name: "subreddit",
  initialState,
  reducers: {
    setCommunityData: (state, action) => {
      state.communityData = action.payload;
    },
    toggleCommunityModalState: (state) => {
      state.communityModalState = !state.communityModalState;
    },
    setUserCommunities: (state, action) => {
      state.userCommunities = action.payload;
    },
  },
});

export const {
  setCommunityData,
  toggleCommunityModalState,
  setUserCommunities,
} = subredditSlice.actions;

export default subredditSlice.reducer;
