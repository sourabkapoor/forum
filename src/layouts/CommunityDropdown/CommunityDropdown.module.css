@font-face {
  font-family: "Open Sans";
  src: url(../../assets/OpenSans-Regular.ttf);
}

.community__dropdown-my-communities {
  font-family: "Open Sans";
  color: rgb(91, 93, 96);
  font-size: 10px;
  padding-left: 0.7rem;
  margin-bottom: 6px;
}
.community__icon {
  width: 20px;
  height: 20px;
}

.community__dropdown-button {
  padding: 6.4px 12.8px;
  background-color: inherit;
  width: 100%;
  box-sizing: border-box;
  word-wrap: break-word;
  display: flex;
  align-items: center;
  color: black;
  gap: 8px;
  border: none;
}
.community__dropdown-button:hover {
  background-color: #e4e4e4;
  cursor: pointer;
  border-radius: 4px;
}

.community-dropdown__link {
  background-color: inherit;
  padding: 8px 24px;
  color: black;
  display: flex;
  align-items: center;
  cursor: pointer;
}

.community-dropdown__link:hover {
  background-color: #272729;
}

.community-dropdown__image {
  height: 20px;
  width: 20px;
  border-radius: 100%;
}

.community-dropdown__community {
  flex-grow: 1;
  margin-left: 8px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  font-size: 14px;
  line-height: 18px;
}
