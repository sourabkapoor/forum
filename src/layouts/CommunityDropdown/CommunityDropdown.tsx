import CSSModules from "react-css-modules";
import { useAppDispatch, useAppSelector } from "../../hooks/hooks";
import Dropdown from "../Dropdown/Dropdown";
import styles from "./CommunityDropdown.module.css";
import {
  selectCommunityData,
  selectUserCommunitiesData,
  setUserCommunities,
  toggleCommunityModalState,
} from "../../features/subreddit/subredditSlice";
import { AiOutlinePlus } from "react-icons/ai";
import { useDeferredValue, useEffect, useReducer, useState } from "react";
import {
  collection,
  doc,
  DocumentData,
  getDoc,
  getDocs,
  onSnapshot,
} from "firebase/firestore";
import { db, getUserId } from "../../firebase";
import { Action } from "@reduxjs/toolkit";
import { Link } from "react-router-dom";

interface Props {
  dropdown: string;
  handleHomeDropdown: Function;
}

const CommunityDropdown: React.FC<Props> = ({
  dropdown,
  handleHomeDropdown,
}) => {
  const { name } = useAppSelector(selectCommunityData);
  const userCommunities = useAppSelector(selectUserCommunitiesData);

  const dispatch = useAppDispatch();

  function toggleCommunity() {
    dispatch(toggleCommunityModalState());
  }

  function toggleDropdown() {
    handleHomeDropdown();
  }

  useEffect(() => {
    const userCommunities = collection(
      db,
      `users/${getUserId()}/communitySnippets`
    );

    onSnapshot(userCommunities, (snapshot) => {
      dispatch(setUserCommunities(snapshot.docs));
    });
  }, [name, dispatch]);

  return (
    <Dropdown dropdown={dropdown}>
      <div styleName="community__dropdown">
        <p styleName="community__dropdown-my-communities">MY COMMUNITIES</p>
        <button
          styleName="community__dropdown-button"
          onClick={toggleCommunity}
        >
          <AiOutlinePlus styleName="community__dropdown-icon" />
          Create Community
        </button>
      </div>
      <div>
        {userCommunities?.map((doc: DocumentData) => (
          <Link
            onClick={toggleDropdown}
            styleName="community-dropdown__link"
            to={`/r/${doc.data().communityId}`}
          >
            <img
              styleName="community-dropdown__image"
              src="https://styles.redditmedia.com/t5_2r0ij/styles/communityIcon_yor9myhxz5x11.png"
              alt="default reddit subreddit logo"
            />
            <span styleName="community-dropdown__community">
              r/{doc.data().communityId}
            </span>
          </Link>
        ))}
      </div>
    </Dropdown>
  );
};

export default CSSModules(CommunityDropdown, styles, {
  allowMultiple: true,
  handleNotFoundStyleName: "log",
});
