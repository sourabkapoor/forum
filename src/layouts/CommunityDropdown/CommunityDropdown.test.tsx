import {
  queryByPlaceholderText,
  queryByText,
  render,
  screen,
} from "@testing-library/react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";
import each from "jest-each";
import CommunityDropdown from "./CommunityDropdown";
import * as subredditSlice from "../../features/subreddit/subredditSlice";
import StoreProvider from "../../redux/provider";

describe("CommunityDropdown", () => {
  test("snapshot", () => {
    const { asFragment } = render(
      <StoreProvider>
        <CommunityDropdown dropdown="community" handleHomeDropdown={jest.fn()}/>
      </StoreProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
  test("opens community modal when user clicks", async () => {
    render(
      <StoreProvider>
        <CommunityDropdown dropdown="community" handleHomeDropdown={jest.fn()}/>
      </StoreProvider>
    );

    // const mockToggleCommunityModalState =
    //   toggleCommunityModalState as jest.MockedFunction<
    //     typeof toggleCommunityModalState
    //   >;

    jest.spyOn(subredditSlice, "toggleCommunityModalState");

    const createCommunityButton = screen.queryByRole("button", {
      name: /Create Community/i,
    });

    const user = userEvent.setup();

    await user.click(createCommunityButton!);

    expect(subredditSlice.toggleCommunityModalState).toHaveBeenCalledTimes(1);
  });
});
