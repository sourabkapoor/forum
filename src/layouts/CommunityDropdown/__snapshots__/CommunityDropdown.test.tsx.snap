// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommunityDropdown snapshot 1`] = `
<DocumentFragment>
  <div
    class="dropdown community-width"
  >
    <div
      class="dropdown__content"
    >
      <div
        class="community__dropdown"
      >
        <p
          class="community__dropdown-my-communities"
        >
          MY COMMUNITIES
        </p>
        <button
          class="community__dropdown-button"
        >
          <svg
            class="community__dropdown-icon"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            t="1551322312294"
            version="1.1"
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <defs />
            <path
              d="M474 152m8 0l60 0q8 0 8 8l0 704q0 8-8 8l-60 0q-8 0-8-8l0-704q0-8 8-8Z"
            />
            <path
              d="M168 474m8 0l672 0q8 0 8 8l0 60q0 8-8 8l-672 0q-8 0-8-8l0-60q0-8 8-8Z"
            />
          </svg>
          Create Community
        </button>
      </div>
      <div />
    </div>
  </div>
</DocumentFragment>
`;
